socialhome/streams/app/stores/streamStore.js

```diff
@@ -57,6 +58,14 @@ function fetchSharesSuccess(state, payload) {
     })
 }
 
+function fetchNewContentSuccess(state, payload) {
+    // It's important content is added before it's id is appended to `contentIds`
+    // to prevent it to be rendered when it's not defined yet
+    Vue.set(state.contents, payload.data.id, payload.data)
+    state.contentIds.unshift(payload.data.id)
```

Comment:
Hmm... I didn't catch it but we may have a problem with this method. Content is added in the order it it fetched, not in the order id were received. Let's say `unfetchedContentIds` is `[34, 36, 39, 73]`. If id 73 is fetched before the other, we may end up with content being added in order `[73, 34, 36, 39]` to the top of the stream. Is it a bug or is it ok ?

Reply:
Hmm... No, it is a bug. In the case of an empty stream, it modifies the logic of fetching more content as `last_id` should be 73 and not 39. Which means 73 would be fetched again.

Reply:
Well, `last_id` should always be the smallest `through` in the stream, because when we scroll we're scrolling backwards in time. So I don't think that applies here as the push notifications for new content will always be newer than anything currently in the stream (= higher `through` id's).

But still, how it *should* work, and how it works currently, is that newest content is at the top. Say you pull in 30 items to the top of the stream. They should be time ordered so that the newest ones are at the top.

I'm actually thinking we should probably think of not pulling new content using the Content API but the stream API. Something like the following:

* Instead of pushing "there is new content X" via websockets and storing a list of unfetched ID's, we would just store a "amount of new content on server", which would be updated by server.
* To pull new content (= user clicks the badge), we do an inverse fetch to the backend, so not `last_id` but `first_id` or something. This makes the server return everything (or X items) *above* the `first_id`.

How would this sound? I think it's doable on the backend, not sure how easily though. It would be better performance for frontend fetching and would easily solve the ordering issue, since the backend would return content in the right order, just like the load more functionality does.

Of course we don't have this backend API, so maybe just go with the current solution for now and pull in content according to the ID's, ignoring the ordering (and writing too many tests). How does that sound?

Reply:
Hmm... Don't know. I always think about bad network coverage when it comes to fetching content. As much as possible, I prefer fetching atomic content one by one instead of one big request. It removes the single point of failure. If one request fail, at least you have the other content. But I admit it comes with nasty algorithmic problems. We could get back to the getter. Though I have no idea what performance cost it has. I don't know what to do :(

Reply:
It's cheaper to fetch a list of content in one go, just like you do when initially loading the stream. There is a lot of overhead doing separate requests. There is also a lot of overhead processing the payload in Vue, especially with older phones.

If the request fails, the client should retry a certain amount of times or at least tell the user to try in a bit.

Of course we shouldn't fetch hundreds of items. I propose to fetch X amount so one can kind of "scroll the stream" upwards. That would actually be quite unique? :) I think some Twitter clients I've used have that option.

Reply:
New code fixes this behaviour.